File: 05-Factory-Method/src/cpp/main.cpp
// Chapter 05 — Factory Method Pattern in C++
// Compile: g++ -std=c++17 -o main main.cpp && ./main

#include <iostream>
#include <vector>
#include <memory>
#include "LogisticsFactory.h"

int main() {
    std::cout << "=== Road Logistics ===\n";
    RoadLogistics road;
    road.planDelivery();

    std::cout << "\n=== Sea Logistics ===\n";
    SeaLogistics sea;
    sea.planDelivery();

    std::cout << "\n=== Air Logistics ===\n";
    AirLogistics air;
    air.planDelivery();

    std::cout << "\n=== All Logistics (polymorphism) ===\n";
    std::vector<const LogisticsFactory*> factories = {&road, &sea, &air};
    for (const auto* f : factories) {
        f->planDelivery();
    }

    return 0;
}

File: 05-Factory-Method/src/go/main.go
// Chapter 05 — Factory Method Pattern in Go
// Run: go run .

package main

import "fmt"

func main() {
	fmt.Println("=== Road Logistics ===")
	PlanDelivery(RoadLogistics{})

	fmt.Println("\n=== Sea Logistics ===")
	PlanDelivery(SeaLogistics{})

	fmt.Println("\n=== Air Logistics ===")
	PlanDelivery(AirLogistics{})

	fmt.Println("\n=== All Logistics (interface polymorphism) ===")
	factories := []LogisticsFactory{RoadLogistics{}, SeaLogistics{}, AirLogistics{}}
	for _, f := range factories {
		PlanDelivery(f)
	}
}
